Import neccessary libraries

In [26]:
import yaml
from pathlib import Path
import logging

import pandas as pd
import numpy as np
import optuna
import mlflow

from sklearn.ensemble import RandomForestClassifier
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from catboost import CatBoostClassifier

from sklearn.metrics import average_precision_score, log_loss, roc_auc_score
from sklearn.model_selection import StratifiedKFold, cross_val_predict
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

# 1. THIẾT LẬP LOGGING (Dùng force=True để Jupyter không bị lỗi khi chạy lại cell nhiều lần)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    force=True
)
logger = logging.getLogger(__name__)

# 2. ĐƯỜNG DẪN DỮ LIỆU
LINEAR_DATA_PATH = Path("../data/processed/cleaned_data_linear.csv")
TREE_DATA_PATH = Path("../data/processed/cleaned_data_tree.csv")

CONFIG_PATH = Path("../configs/models/best_model_config.yaml")
TARGET_COLUMN = "churn_status"
SPLIT_COLUMN = "data_split"
RANDOM_STATE = 42

In [27]:
mlflow.set_tracking_uri("http://127.0.0.1:5000/")
mlflow.set_experiment("Telcom_All_Models_Tuning")
# autolog can create many extra logs in notebook runs
# Keep it off for cleaner manual logging
# mlflow.sklearn.autolog()

overall_results = {}

In [28]:
from sklearn.model_selection import train_test_split

def validate_and_split(
    df: pd.DataFrame,
    dataset_name: str,
    val_size: float = 0.2,
    random_state: int = 42,
):
    required_cols = {TARGET_COLUMN, SPLIT_COLUMN}
    missing_cols = required_cols - set(df.columns)
    if missing_cols:
        raise ValueError(f"[{dataset_name}] Missing required columns: {missing_cols}")

    split_values = set(df[SPLIT_COLUMN].dropna().unique())
    if not {"train", "test"}.issubset(split_values):
        raise ValueError(
            f"[{dataset_name}] Column '{SPLIT_COLUMN}' must contain both 'train' and 'test'. "
            f"Found: {split_values}"
        )

    train_set = df[df[SPLIT_COLUMN] == "train"].copy()
    test_set = df[df[SPLIT_COLUMN] == "test"].copy()

    if train_set.empty:
        raise ValueError(f"[{dataset_name}] Train set is empty.")
    if test_set.empty:
        raise ValueError(f"[{dataset_name}] Test set is empty.")

    if train_set[TARGET_COLUMN].nunique() < 2:
        raise ValueError(f"[{dataset_name}] Train set must contain at least 2 target classes.")

    # Split train -> train_sub + val_set
    train_sub, val_set = train_test_split(
        train_set,
        test_size=val_size,
        random_state=random_state,
        stratify=train_set[TARGET_COLUMN],
    )

    # Train
    X_train = train_sub.drop(columns=[TARGET_COLUMN, SPLIT_COLUMN])
    y_train = train_sub[TARGET_COLUMN]

    # Validation
    X_val = val_set.drop(columns=[TARGET_COLUMN, SPLIT_COLUMN])
    y_val = val_set[TARGET_COLUMN]

    # Test
    X_test = test_set.drop(columns=[TARGET_COLUMN, SPLIT_COLUMN])
    y_test = test_set[TARGET_COLUMN]

    return X_train, X_val, X_test, y_train, y_val, y_test

In [29]:
df_linear = pd.read_csv(LINEAR_DATA_PATH)
df_tree = pd.read_csv(TREE_DATA_PATH)

logger.info("Loaded linear dataset: %s | shape=%s", LINEAR_DATA_PATH, df_linear.shape)
logger.info("Loaded tree dataset: %s | shape=%s", TREE_DATA_PATH, df_tree.shape)

X_train_linear, X_val_linear, X_test_linear, y_train_linear, y_val_linear, y_test_linear = validate_and_split(
    df_linear, "linear", val_size=0.2, random_state=RANDOM_STATE
)

X_train_tree, X_val_tree, X_test_tree, y_train_tree, y_val_tree, y_test_tree = validate_and_split(
    df_tree, "tree", val_size=0.2, random_state=RANDOM_STATE
)

logger.info(
    "Linear split shapes | train=%s val=%s test=%s",
    X_train_linear.shape,
    X_val_linear.shape,
    X_test_linear.shape,
)
logger.info(
    "Tree split shapes | train=%s val=%s test=%s",
    X_train_tree.shape,
    X_val_tree.shape,
    X_test_tree.shape,
)

cv_linear = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
cv_tree = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)

2026-04-15 15:41:38 [INFO] Loaded linear dataset: ..\data\processed\cleaned_data_linear.csv | shape=(7043, 51)
2026-04-15 15:41:38 [INFO] Loaded tree dataset: ..\data\processed\cleaned_data_tree.csv | shape=(7043, 51)
2026-04-15 15:41:38 [INFO] Linear split shapes | train=(4507, 49) val=(1127, 49) test=(1409, 49)
2026-04-15 15:41:38 [INFO] Tree split shapes | train=(4507, 49) val=(1127, 49) test=(1409, 49)


### BASELINE

In [30]:
from sklearn.linear_model import LogisticRegression

with mlflow.start_run(run_name="Baseline_LogisticRegression"):
    baseline_model = LogisticRegression(
        random_state=RANDOM_STATE,
        class_weight="balanced",
        max_iter=1000
    )
    baseline_model.fit(X_train_linear, y_train_linear)

    # Evaluate on validation set, not test set
    y_val_proba = baseline_model.predict_proba(X_val_linear)[:, 1]
    y_val_pred = (y_val_proba >= 0.5).astype(int)

    val_metrics = {
        "val_pr_auc": float(average_precision_score(y_val_linear, y_val_proba)),
        "val_roc_auc": float(roc_auc_score(y_val_linear, y_val_proba)),
        "val_log_loss": float(log_loss(y_val_linear, y_val_proba)),
        "val_accuracy": float(accuracy_score(y_val_linear, y_val_pred)),
        "val_precision": float(precision_score(y_val_linear, y_val_pred, zero_division=0)),
        "val_recall": float(recall_score(y_val_linear, y_val_pred, zero_division=0)),
        "val_f1_score": float(f1_score(y_val_linear, y_val_pred, zero_division=0)),
        "val_predicted_positive_rate": float(y_val_pred.mean()),
    }

    mlflow.log_params({
        "model_type": "LogisticRegression",
        "dataset_type": "linear",
        "class_weight": "balanced",
        "max_iter": 1000,
        "random_state": RANDOM_STATE,
        "default_threshold": 0.5,
    })
    mlflow.log_metrics(val_metrics)

    overall_results["Baseline_LogisticRegression"] = {
        "model": baseline_model,
        "dataset_type": "linear",
        "cv_pr_auc": None,
        "val_pr_auc": float(val_metrics["val_pr_auc"]),
        "val_roc_auc": float(val_metrics["val_roc_auc"]),
        "val_log_loss": float(val_metrics["val_log_loss"]),
        "accuracy": float(val_metrics["val_accuracy"]),
        "precision": float(val_metrics["val_precision"]),
        "recall": float(val_metrics["val_recall"]),
        "f1_score": float(val_metrics["val_f1_score"]),
        "predicted_positive_rate": float(val_metrics["val_predicted_positive_rate"]),
        "threshold": 0.5,
        "params": {
            "class_weight": "balanced",
            "max_iter": 1000,
            "random_state": RANDOM_STATE,
        },
    }

🏃 View run Baseline_LogisticRegression at: http://127.0.0.1:5000/#/experiments/3/runs/faf53db4f4784ef18e2a9dd21cb2b428
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


### RANDOM FOREST

In [31]:
def objective_rf(trial):
    """
    Tune RandomForest using cross-validated PR-AUC on the training set only.
    This avoids validation/test leakage during hyperparameter search.
    """
    params = {
        "n_estimators": trial.suggest_int("n_estimators", 100, 500),
        "max_depth": trial.suggest_int("max_depth", 5, 25),
        "min_samples_split": trial.suggest_int("min_samples_split", 2, 10),
        "min_samples_leaf": trial.suggest_int("min_samples_leaf", 1, 5),
        "criterion": trial.suggest_categorical("criterion", ["gini", "entropy"]),
        "random_state": RANDOM_STATE,
        "class_weight": "balanced",
        "n_jobs": -1,
    }

    model = RandomForestClassifier(**params)

    y_proba_oof = cross_val_predict(
        estimator=model,
        X=X_train_tree,
        y=y_train_tree,
        cv=cv_tree,
        method="predict_proba",
        n_jobs=-1,
    )[:, 1]

    y_pred_oof = (y_proba_oof >= 0.5).astype(int)

    metrics = {
        "cv_pr_auc": float(average_precision_score(y_train_tree, y_proba_oof)),
        "cv_roc_auc": float(roc_auc_score(y_train_tree, y_proba_oof)),
        "cv_log_loss": float(log_loss(y_train_tree, y_proba_oof)),
        "cv_accuracy": float(accuracy_score(y_train_tree, y_pred_oof)),
        "cv_precision": float(precision_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_recall": float(recall_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_f1_score": float(f1_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_predicted_positive_rate": float(y_pred_oof.mean()),
    }

    with mlflow.start_run(nested=True):
        mlflow.log_params(params)
        mlflow.log_metrics(metrics)

    return metrics["cv_pr_auc"]


with mlflow.start_run(run_name="RandomForest_PR_AUC_Optimization_CV"):
    logger.info("Starting RandomForest hyperparameter tuning with Stratified CV...")

    study_rf = optuna.create_study(direction="maximize")
    study_rf.optimize(objective_rf, n_trials=10)

    mlflow.log_params(study_rf.best_params)
    mlflow.log_metric("best_cv_pr_auc", float(study_rf.best_value))

    best_rf = RandomForestClassifier(
        **study_rf.best_params,
        random_state=RANDOM_STATE,
        class_weight="balanced",
        n_jobs=-1,
    )
    best_rf.fit(X_train_tree, y_train_tree)

    # Evaluate on validation set only at this stage
    y_val_proba = best_rf.predict_proba(X_val_tree)[:, 1]
    y_val_pred = (y_val_proba >= 0.5).astype(int)

    val_metrics = {
        "val_pr_auc": float(average_precision_score(y_val_tree, y_val_proba)),
        "val_roc_auc": float(roc_auc_score(y_val_tree, y_val_proba)),
        "val_log_loss": float(log_loss(y_val_tree, y_val_proba)),
        "val_accuracy": float(accuracy_score(y_val_tree, y_val_pred)),
        "val_precision": float(precision_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_recall": float(recall_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_f1_score": float(f1_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_predicted_positive_rate": float(y_val_pred.mean()),
    }

    mlflow.log_params({
        "dataset_type": "tree",
        "default_threshold": 0.5,
    })
    mlflow.log_metrics(val_metrics)
    mlflow.sklearn.log_model(best_rf, name="best_rf_prauc_model")

    overall_results["RandomForest"] = {
        "model": best_rf,
        "dataset_type": "tree",
        "cv_pr_auc": float(study_rf.best_value),
        "val_pr_auc": float(val_metrics["val_pr_auc"]),
        "val_roc_auc": float(val_metrics["val_roc_auc"]),
        "val_log_loss": float(val_metrics["val_log_loss"]),
        "accuracy": float(val_metrics["val_accuracy"]),
        "precision": float(val_metrics["val_precision"]),
        "recall": float(val_metrics["val_recall"]),
        "f1_score": float(val_metrics["val_f1_score"]),
        "predicted_positive_rate": float(val_metrics["val_predicted_positive_rate"]),
        "threshold": 0.5,
        "params": study_rf.best_params,
    }

    print("-" * 40)
    print("RandomForest tuning completed!")
    print(f"Best CV PR-AUC: {study_rf.best_value:.4f}")
    print(f"Validation PR-AUC: {val_metrics['val_pr_auc']:.4f}")
    print(f"Best Parameters: {study_rf.best_params}")

2026-04-15 15:48:08 [INFO] Starting RandomForest hyperparameter tuning with Stratified CV...
[I 2026-04-15 15:48:08,994] A new study created in memory with name: no-name-4061a9d8-8ce0-42c3-aad3-75c7a131ed6f
[I 2026-04-15 15:48:16,920] Trial 0 finished with value: 0.6525970273442889 and parameters: {'n_estimators': 350, 'max_depth': 23, 'min_samples_split': 3, 'min_samples_leaf': 4, 'criterion': 'gini'}. Best is trial 0 with value: 0.6525970273442889.


🏃 View run dapper-vole-435 at: http://127.0.0.1:5000/#/experiments/3/runs/7f3a4f112dea4c1999841f8b3d9da0ee
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:22,811] Trial 1 finished with value: 0.6421281999185545 and parameters: {'n_estimators': 127, 'max_depth': 20, 'min_samples_split': 7, 'min_samples_leaf': 1, 'criterion': 'gini'}. Best is trial 0 with value: 0.6525970273442889.


🏃 View run victorious-hog-644 at: http://127.0.0.1:5000/#/experiments/3/runs/f9620182a19e438893a655c2ca9049f8
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:30,862] Trial 2 finished with value: 0.6519727679063567 and parameters: {'n_estimators': 488, 'max_depth': 13, 'min_samples_split': 7, 'min_samples_leaf': 3, 'criterion': 'gini'}. Best is trial 0 with value: 0.6525970273442889.


🏃 View run bemused-trout-614 at: http://127.0.0.1:5000/#/experiments/3/runs/271d231817d6469882e742d85f10515b
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:32,197] Trial 3 finished with value: 0.6543825643013963 and parameters: {'n_estimators': 199, 'max_depth': 11, 'min_samples_split': 6, 'min_samples_leaf': 5, 'criterion': 'entropy'}. Best is trial 3 with value: 0.6543825643013963.


🏃 View run luminous-kit-605 at: http://127.0.0.1:5000/#/experiments/3/runs/29807fcb48404a768a0eb9f6a8053004
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:34,444] Trial 4 finished with value: 0.6519993304822099 and parameters: {'n_estimators': 389, 'max_depth': 17, 'min_samples_split': 7, 'min_samples_leaf': 5, 'criterion': 'gini'}. Best is trial 3 with value: 0.6543825643013963.


🏃 View run casual-skunk-560 at: http://127.0.0.1:5000/#/experiments/3/runs/3b40f9f42b33438b9a4838845dc03ed8
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:37,292] Trial 5 finished with value: 0.6500189956084924 and parameters: {'n_estimators': 457, 'max_depth': 12, 'min_samples_split': 6, 'min_samples_leaf': 1, 'criterion': 'gini'}. Best is trial 3 with value: 0.6543825643013963.


🏃 View run masked-dolphin-726 at: http://127.0.0.1:5000/#/experiments/3/runs/51cb98c2c5f64a7eb0f38296c00b2be6
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:39,072] Trial 6 finished with value: 0.654435649370918 and parameters: {'n_estimators': 321, 'max_depth': 6, 'min_samples_split': 9, 'min_samples_leaf': 4, 'criterion': 'gini'}. Best is trial 6 with value: 0.654435649370918.


🏃 View run fun-whale-975 at: http://127.0.0.1:5000/#/experiments/3/runs/4517fc95f78c4f4e942a04ba006ec660
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:40,721] Trial 7 finished with value: 0.6548677188320302 and parameters: {'n_estimators': 282, 'max_depth': 10, 'min_samples_split': 5, 'min_samples_leaf': 3, 'criterion': 'gini'}. Best is trial 7 with value: 0.6548677188320302.


🏃 View run industrious-kit-921 at: http://127.0.0.1:5000/#/experiments/3/runs/6aa810d163e247a7bedd8c025d96b088
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:42,240] Trial 8 finished with value: 0.650378054954975 and parameters: {'n_estimators': 214, 'max_depth': 18, 'min_samples_split': 6, 'min_samples_leaf': 3, 'criterion': 'entropy'}. Best is trial 7 with value: 0.6548677188320302.


🏃 View run burly-gull-298 at: http://127.0.0.1:5000/#/experiments/3/runs/59b931a68d2240e08f5af1ec43ebd952
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:48:44,327] Trial 9 finished with value: 0.6589818616283536 and parameters: {'n_estimators': 344, 'max_depth': 9, 'min_samples_split': 10, 'min_samples_leaf': 4, 'criterion': 'gini'}. Best is trial 9 with value: 0.6589818616283536.


🏃 View run fearless-carp-116 at: http://127.0.0.1:5000/#/experiments/3/runs/72655e4df3ec445b88682900b5cfda20
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


2026/04/15 15:48:45 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


----------------------------------------
RandomForest tuning completed!
Best CV PR-AUC: 0.6590
Validation PR-AUC: 0.6732
Best Parameters: {'n_estimators': 344, 'max_depth': 9, 'min_samples_split': 10, 'min_samples_leaf': 4, 'criterion': 'gini'}
🏃 View run RandomForest_PR_AUC_Optimization_CV at: http://127.0.0.1:5000/#/experiments/3/runs/cf6873106752482a8bd68f1c702ee623
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


In [ ]:
# import matplotlib.pyplot as plt
# import pandas as pd

# # 1. Lấy độ quan trọng của các feature từ model tốt nhất
# importances = best_model.feature_importances_
# feature_names = X_train.columns

# # 2. Tạo DataFrame để dễ sắp xếp
# feature_importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': importances})
# feature_importance_df = feature_importance_df.sort_values(by='Importance', ascending=False)

# # 3. In ra Top 5
# print("Top 5 Important Features:")
# print(feature_importance_df.head(5))

# # 4. Vẽ biểu đồ trực quan
# plt.figure(figsize=(10, 6))
# plt.barh(feature_importance_df['Feature'].head(5), feature_importance_df['Importance'].head(5), color='skyblue')
# plt.xlabel('Importance Score')
# plt.title('Top 5 Features causing potential Data Leakage')
# plt.gca().invert_yaxis()
# plt.show()

### XGBOOST

In [32]:
# =========================================
# XGBoost tuning with CV
# =========================================

# Class imbalance ratio for positive class weighting
ratio_tree = float(y_train_tree.value_counts()[0] / y_train_tree.value_counts()[1])

def objective_xgb(trial):
    """
    Tune XGBoost using cross-validated PR-AUC on the training set only.
    This avoids validation/test leakage during hyperparameter search.
    """
    params = {
        "n_estimators": trial.suggest_int("n_estimators", 100, 1000),
        "max_depth": trial.suggest_int("max_depth", 5, 15),
        "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3, log=True),
        "subsample": trial.suggest_float("subsample", 0.5, 1.0),
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.5, 1.0),
        "gamma": trial.suggest_float("gamma", 0.0, 5.0),
        "scale_pos_weight": trial.suggest_float("scale_pos_weight", 1.0, ratio_tree * 1.5),
        "random_state": RANDOM_STATE,
        "eval_metric": "logloss",
        "n_jobs": -1,
    }

    model = XGBClassifier(**params)

    y_proba_oof = cross_val_predict(
        estimator=model,
        X=X_train_tree,
        y=y_train_tree,
        cv=cv_tree,
        method="predict_proba",
        n_jobs=-1,
    )[:, 1]

    y_pred_oof = (y_proba_oof >= 0.5).astype(int)

    metrics = {
        "cv_pr_auc": float(average_precision_score(y_train_tree, y_proba_oof)),
        "cv_roc_auc": float(roc_auc_score(y_train_tree, y_proba_oof)),
        "cv_log_loss": float(log_loss(y_train_tree, y_proba_oof)),
        "cv_accuracy": float(accuracy_score(y_train_tree, y_pred_oof)),
        "cv_precision": float(precision_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_recall": float(recall_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_f1_score": float(f1_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_predicted_positive_rate": float(y_pred_oof.mean()),
    }

    with mlflow.start_run(nested=True):
        mlflow.log_params(params)
        mlflow.log_metrics(metrics)

    return metrics["cv_pr_auc"]


with mlflow.start_run(run_name="XGBoost_PR_AUC_Optimization_CV"):
    logger.info("Starting XGBoost hyperparameter tuning with Stratified CV...")

    study_xgb = optuna.create_study(direction="maximize")
    study_xgb.optimize(objective_xgb, n_trials=50)

    mlflow.log_params(study_xgb.best_params)
    mlflow.log_metric("best_cv_pr_auc", float(study_xgb.best_value))
    mlflow.log_params({
        "dataset_type": "tree",
        "default_threshold": 0.5,
    })

    best_xgb = XGBClassifier(
        **study_xgb.best_params,
        random_state=RANDOM_STATE,
        eval_metric="logloss",
        n_jobs=-1,
    )
    best_xgb.fit(X_train_tree, y_train_tree)

    # Evaluate on validation set only at this stage
    y_val_proba = best_xgb.predict_proba(X_val_tree)[:, 1]
    y_val_pred = (y_val_proba >= 0.5).astype(int)

    val_metrics = {
        "val_pr_auc": float(average_precision_score(y_val_tree, y_val_proba)),
        "val_roc_auc": float(roc_auc_score(y_val_tree, y_val_proba)),
        "val_log_loss": float(log_loss(y_val_tree, y_val_proba)),
        "val_accuracy": float(accuracy_score(y_val_tree, y_val_pred)),
        "val_precision": float(precision_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_recall": float(recall_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_f1_score": float(f1_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_predicted_positive_rate": float(y_val_pred.mean()),
    }

    mlflow.log_metrics(val_metrics)
    mlflow.xgboost.log_model(best_xgb, name="best_xgb_prauc_model")

    overall_results["XGBoost"] = {
        "model": best_xgb,
        "dataset_type": "tree",
        "cv_pr_auc": float(study_xgb.best_value),
        "val_pr_auc": float(val_metrics["val_pr_auc"]),
        "val_roc_auc": float(val_metrics["val_roc_auc"]),
        "val_log_loss": float(val_metrics["val_log_loss"]),
        "accuracy": float(val_metrics["val_accuracy"]),
        "precision": float(val_metrics["val_precision"]),
        "recall": float(val_metrics["val_recall"]),
        "f1_score": float(val_metrics["val_f1_score"]),
        "predicted_positive_rate": float(val_metrics["val_predicted_positive_rate"]),
        "threshold": 0.5,
        "params": study_xgb.best_params,
    }

    print("-" * 40)
    print("XGBoost tuning completed!")
    print(f"Best CV PR-AUC: {study_xgb.best_value:.4f}")
    print(f"Validation PR-AUC: {val_metrics['val_pr_auc']:.4f}")
    print(f"Best Parameters: {study_xgb.best_params}")

2026-04-15 15:49:20 [INFO] Starting XGBoost hyperparameter tuning with Stratified CV...
[I 2026-04-15 15:49:20,135] A new study created in memory with name: no-name-4e6843eb-8760-4362-92c2-ac79d7c3f1ae
[I 2026-04-15 15:49:21,204] Trial 0 finished with value: 0.6168803636112163 and parameters: {'n_estimators': 371, 'max_depth': 8, 'learning_rate': 0.1753088361370108, 'subsample': 0.5345356543620612, 'colsample_bytree': 0.6848683688297144, 'gamma': 1.3748225242012606, 'scale_pos_weight': 1.2002450587743934}. Best is trial 0 with value: 0.6168803636112163.


🏃 View run unleashed-flea-797 at: http://127.0.0.1:5000/#/experiments/3/runs/2b3a1d5c7420483cb068cfaf4ceeaab1
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:22,953] Trial 1 finished with value: 0.6482617286319211 and parameters: {'n_estimators': 315, 'max_depth': 11, 'learning_rate': 0.014368007695185628, 'subsample': 0.6017378080904339, 'colsample_bytree': 0.7118835897050821, 'gamma': 1.2309978684095146, 'scale_pos_weight': 1.4475302842413424}. Best is trial 1 with value: 0.6482617286319211.


🏃 View run awesome-boar-873 at: http://127.0.0.1:5000/#/experiments/3/runs/b108f6863c1e4b60a1fa6df972411523
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:23,779] Trial 2 finished with value: 0.6491951480874624 and parameters: {'n_estimators': 359, 'max_depth': 15, 'learning_rate': 0.218993437189774, 'subsample': 0.6981675573336185, 'colsample_bytree': 0.9825036249761587, 'gamma': 4.41683146967094, 'scale_pos_weight': 1.8287269394243466}. Best is trial 2 with value: 0.6491951480874624.


🏃 View run chill-rook-450 at: http://127.0.0.1:5000/#/experiments/3/runs/7d09586801b54c66b4646b82c3ea4742
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:25,193] Trial 3 finished with value: 0.5975377912248826 and parameters: {'n_estimators': 460, 'max_depth': 11, 'learning_rate': 0.15767129173311661, 'subsample': 0.6640777528260788, 'colsample_bytree': 0.7986062478856284, 'gamma': 0.5629251854984485, 'scale_pos_weight': 1.1476365461852978}. Best is trial 2 with value: 0.6491951480874624.


🏃 View run receptive-goose-940 at: http://127.0.0.1:5000/#/experiments/3/runs/06e56fed8adf4ee6bfd90e9bfa694180
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:25,812] Trial 4 finished with value: 0.6444862419821613 and parameters: {'n_estimators': 205, 'max_depth': 9, 'learning_rate': 0.2958049393294869, 'subsample': 0.9223557041178339, 'colsample_bytree': 0.7819967327687962, 'gamma': 3.265596799780768, 'scale_pos_weight': 2.1322195393851837}. Best is trial 2 with value: 0.6491951480874624.


🏃 View run bedecked-ape-343 at: http://127.0.0.1:5000/#/experiments/3/runs/f2d121fcc97c4ad6bed07cc4cfe95b88
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:27,292] Trial 5 finished with value: 0.6325072736010999 and parameters: {'n_estimators': 393, 'max_depth': 15, 'learning_rate': 0.03914833163562197, 'subsample': 0.6524229552618552, 'colsample_bytree': 0.523799571748547, 'gamma': 1.074757071697507, 'scale_pos_weight': 3.7203188778967045}. Best is trial 2 with value: 0.6491951480874624.


🏃 View run sedate-cod-497 at: http://127.0.0.1:5000/#/experiments/3/runs/35e7d9f6d80a482991a23dc3ccc646e3
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:28,285] Trial 6 finished with value: 0.642558805878552 and parameters: {'n_estimators': 424, 'max_depth': 10, 'learning_rate': 0.15679739707733228, 'subsample': 0.6553521541015614, 'colsample_bytree': 0.7661314109336752, 'gamma': 4.542494535555362, 'scale_pos_weight': 2.748050120666674}. Best is trial 2 with value: 0.6491951480874624.


🏃 View run colorful-chimp-947 at: http://127.0.0.1:5000/#/experiments/3/runs/86b78ef66047422397238e9f5c4e2fe1
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:31,188] Trial 7 finished with value: 0.6382376897122781 and parameters: {'n_estimators': 453, 'max_depth': 11, 'learning_rate': 0.012152055427076772, 'subsample': 0.6063638468520367, 'colsample_bytree': 0.9395676919689646, 'gamma': 0.6398547874896271, 'scale_pos_weight': 2.7773346766435236}. Best is trial 2 with value: 0.6491951480874624.


🏃 View run magnificent-eel-838 at: http://127.0.0.1:5000/#/experiments/3/runs/805fc427ce30478daff4ecdec0419d04
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:32,648] Trial 8 finished with value: 0.6597844479062909 and parameters: {'n_estimators': 795, 'max_depth': 5, 'learning_rate': 0.033515203707455754, 'subsample': 0.7326489053653753, 'colsample_bytree': 0.7570853726247899, 'gamma': 4.843820468088738, 'scale_pos_weight': 2.06940652930305}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run unleashed-sow-850 at: http://127.0.0.1:5000/#/experiments/3/runs/a4ddaf768902496792515ecc4c185319
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:33,535] Trial 9 finished with value: 0.6532505808624075 and parameters: {'n_estimators': 148, 'max_depth': 8, 'learning_rate': 0.027497297504130155, 'subsample': 0.8740192326475811, 'colsample_bytree': 0.946018812423792, 'gamma': 2.659820461323586, 'scale_pos_weight': 2.302874667288039}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run popular-pug-686 at: http://127.0.0.1:5000/#/experiments/3/runs/c534651844b04f4bb29e1b6ac9e8b54e
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:34,988] Trial 10 finished with value: 0.6531111183864209 and parameters: {'n_estimators': 893, 'max_depth': 6, 'learning_rate': 0.0663956302573785, 'subsample': 0.8027626159451884, 'colsample_bytree': 0.575097171691541, 'gamma': 3.35409655197263, 'scale_pos_weight': 3.310812867285283}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run omniscient-bird-89 at: http://127.0.0.1:5000/#/experiments/3/runs/0f19dab84a384164b5eb5a4e953bb17a
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:36,371] Trial 11 finished with value: 0.6530222243107363 and parameters: {'n_estimators': 766, 'max_depth': 5, 'learning_rate': 0.03091108712111805, 'subsample': 0.8412926977533886, 'colsample_bytree': 0.8671873545814202, 'gamma': 2.2835557706774203, 'scale_pos_weight': 2.224797532453476}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run dapper-wren-172 at: http://127.0.0.1:5000/#/experiments/3/runs/5cd7320d5b40472694808e7562b9909b
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:37,652] Trial 12 finished with value: 0.650660469677919 and parameters: {'n_estimators': 668, 'max_depth': 7, 'learning_rate': 0.02192564353784173, 'subsample': 0.9745193394289816, 'colsample_bytree': 0.8860187562968816, 'gamma': 3.416006489612386, 'scale_pos_weight': 2.341269326975534}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run welcoming-quail-938 at: http://127.0.0.1:5000/#/experiments/3/runs/dd6fe6d415674372827b384f568fd7dd
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:38,322] Trial 13 finished with value: 0.6591409985926443 and parameters: {'n_estimators': 103, 'max_depth': 5, 'learning_rate': 0.07192096542704518, 'subsample': 0.7858305546096527, 'colsample_bytree': 0.6381379871470484, 'gamma': 2.4685618925655146, 'scale_pos_weight': 1.718914048838119}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run sassy-bass-201 at: http://127.0.0.1:5000/#/experiments/3/runs/fbcb597825bc44c58db986ce62f4372d
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:39,959] Trial 14 finished with value: 0.6503893772724177 and parameters: {'n_estimators': 982, 'max_depth': 5, 'learning_rate': 0.07351298330717726, 'subsample': 0.7566658531915673, 'colsample_bytree': 0.6252995835202947, 'gamma': 2.180409045781716, 'scale_pos_weight': 1.6615613111193945}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run fortunate-crab-168 at: http://127.0.0.1:5000/#/experiments/3/runs/e6c46aa31a4540b3831c0c3aa7005f51
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:41,102] Trial 15 finished with value: 0.6568950267612443 and parameters: {'n_estimators': 594, 'max_depth': 6, 'learning_rate': 0.09371505140921409, 'subsample': 0.7495907778324935, 'colsample_bytree': 0.6487572159034273, 'gamma': 3.9445401931728403, 'scale_pos_weight': 1.7797881654790646}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run debonair-midge-563 at: http://127.0.0.1:5000/#/experiments/3/runs/1f22cd8517014498a39413d89885f045
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:42,482] Trial 16 finished with value: 0.657571968263461 and parameters: {'n_estimators': 766, 'max_depth': 5, 'learning_rate': 0.04187226632726291, 'subsample': 0.7626843429791487, 'colsample_bytree': 0.5851310448274104, 'gamma': 4.934025119359983, 'scale_pos_weight': 3.1188379377952735}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run skillful-quail-435 at: http://127.0.0.1:5000/#/experiments/3/runs/a0c310f6b5c04ec1991f435808db9fd0
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:43,866] Trial 17 finished with value: 0.6418269742874253 and parameters: {'n_estimators': 764, 'max_depth': 13, 'learning_rate': 0.0960560626807908, 'subsample': 0.8592689812073319, 'colsample_bytree': 0.7051244567919572, 'gamma': 1.8191188380348062, 'scale_pos_weight': 1.9511184155392072}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run crawling-slug-516 at: http://127.0.0.1:5000/#/experiments/3/runs/114d14bf127d4158a91b64a1da4cbbcc
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:44,697] Trial 18 finished with value: 0.6553443389155056 and parameters: {'n_estimators': 103, 'max_depth': 7, 'learning_rate': 0.01960273429241098, 'subsample': 0.503823150516711, 'colsample_bytree': 0.8221340984212818, 'gamma': 0.04016211952449833, 'scale_pos_weight': 1.4419754331280605}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run fortunate-midge-133 at: http://127.0.0.1:5000/#/experiments/3/runs/09c238b389cc4f969e55def2bab005e2
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:45,990] Trial 19 finished with value: 0.6488513102899974 and parameters: {'n_estimators': 571, 'max_depth': 7, 'learning_rate': 0.0497508727855936, 'subsample': 0.7066780575116326, 'colsample_bytree': 0.7347631138096139, 'gamma': 2.913997297036353, 'scale_pos_weight': 2.5178501378084146}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run sassy-trout-328 at: http://127.0.0.1:5000/#/experiments/3/runs/08ffc161a9be42e3b8eeb16cdaca1b61
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:46,796] Trial 20 finished with value: 0.6461081638835874 and parameters: {'n_estimators': 260, 'max_depth': 13, 'learning_rate': 0.10576788496778285, 'subsample': 0.8090721267483839, 'colsample_bytree': 0.6532544825954761, 'gamma': 3.968179057400606, 'scale_pos_weight': 3.930062316243418}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run smiling-jay-300 at: http://127.0.0.1:5000/#/experiments/3/runs/8dc370e5a3e048219ca96913523cdf3b
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:48,336] Trial 21 finished with value: 0.658898746125346 and parameters: {'n_estimators': 803, 'max_depth': 5, 'learning_rate': 0.0526446035538868, 'subsample': 0.7710748157578323, 'colsample_bytree': 0.5920614391492451, 'gamma': 4.814191672211288, 'scale_pos_weight': 3.2332244192619264}. Best is trial 8 with value: 0.6597844479062909.


🏃 View run welcoming-owl-249 at: http://127.0.0.1:5000/#/experiments/3/runs/5689a460f4ba4ef6b006b5766cb290da
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:49,819] Trial 22 finished with value: 0.660352807351118 and parameters: {'n_estimators': 893, 'max_depth': 6, 'learning_rate': 0.06320813698212507, 'subsample': 0.7983302554117204, 'colsample_bytree': 0.5285270822402186, 'gamma': 4.974268066650443, 'scale_pos_weight': 3.234120666708022}. Best is trial 22 with value: 0.660352807351118.


🏃 View run incongruous-colt-623 at: http://127.0.0.1:5000/#/experiments/3/runs/d13e1b4fe9644e53af88d6acdc77d539
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:51,296] Trial 23 finished with value: 0.6562665183903985 and parameters: {'n_estimators': 967, 'max_depth': 6, 'learning_rate': 0.06670317436820027, 'subsample': 0.9095491154317112, 'colsample_bytree': 0.5186349050263942, 'gamma': 4.014584262451917, 'scale_pos_weight': 3.48229248122447}. Best is trial 22 with value: 0.660352807351118.


🏃 View run resilient-owl-187 at: http://127.0.0.1:5000/#/experiments/3/runs/af72d04fc98c48a28c65f9b1f1665bc3
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:52,914] Trial 24 finished with value: 0.6564820693209601 and parameters: {'n_estimators': 877, 'max_depth': 6, 'learning_rate': 0.032820122308269764, 'subsample': 0.7097596249263313, 'colsample_bytree': 0.5015409212488755, 'gamma': 4.199013304303571, 'scale_pos_weight': 2.9078924096754477}. Best is trial 22 with value: 0.660352807351118.


🏃 View run clumsy-lark-983 at: http://127.0.0.1:5000/#/experiments/3/runs/730426334d1e46d8831b67120a897573
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:54,146] Trial 25 finished with value: 0.6499074257990123 and parameters: {'n_estimators': 706, 'max_depth': 8, 'learning_rate': 0.11005518181292602, 'subsample': 0.81339307430406, 'colsample_bytree': 0.5666643808311004, 'gamma': 3.6629801499726207, 'scale_pos_weight': 2.5240447948344995}. Best is trial 22 with value: 0.660352807351118.


🏃 View run flawless-snake-126 at: http://127.0.0.1:5000/#/experiments/3/runs/d4ca3f2d2fd2487da0d21b4c41989b4f
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:55,689] Trial 26 finished with value: 0.6440592582951826 and parameters: {'n_estimators': 848, 'max_depth': 5, 'learning_rate': 0.07731550241477816, 'subsample': 0.7302323204973492, 'colsample_bytree': 0.6232740635946636, 'gamma': 2.8392861732041914, 'scale_pos_weight': 4.118815128602051}. Best is trial 22 with value: 0.660352807351118.


🏃 View run delicate-elk-901 at: http://127.0.0.1:5000/#/experiments/3/runs/22fd48401deb4c1b85c1f6eaf04483b3
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:56,955] Trial 27 finished with value: 0.6513868286032298 and parameters: {'n_estimators': 679, 'max_depth': 7, 'learning_rate': 0.04455273090868656, 'subsample': 0.895661505102807, 'colsample_bytree': 0.5474230509372386, 'gamma': 1.7807008577044625, 'scale_pos_weight': 1.4917262169999297}. Best is trial 22 with value: 0.660352807351118.


🏃 View run defiant-cat-210 at: http://127.0.0.1:5000/#/experiments/3/runs/1a40fc7b5f9d4b30811ce767787565bc
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:58,361] Trial 28 finished with value: 0.6554598757401369 and parameters: {'n_estimators': 918, 'max_depth': 6, 'learning_rate': 0.022824554085180464, 'subsample': 0.9742848738774657, 'colsample_bytree': 0.6716720772756273, 'gamma': 4.510252022268008, 'scale_pos_weight': 1.994373022062727}. Best is trial 22 with value: 0.660352807351118.


🏃 View run youthful-slug-273 at: http://127.0.0.1:5000/#/experiments/3/runs/e6000b75653e4f1fa9592bc55d2083b0
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:49:59,513] Trial 29 finished with value: 0.6562229329434018 and parameters: {'n_estimators': 587, 'max_depth': 8, 'learning_rate': 0.1270670059686334, 'subsample': 0.7920235951169171, 'colsample_bytree': 0.6105106721890963, 'gamma': 4.981852574770069, 'scale_pos_weight': 1.0673677049100159}. Best is trial 22 with value: 0.660352807351118.


🏃 View run painted-robin-994 at: http://127.0.0.1:5000/#/experiments/3/runs/05f8bff7db4643258b0e33c9c60152f8
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:00,906] Trial 30 finished with value: 0.6290757200856391 and parameters: {'n_estimators': 487, 'max_depth': 7, 'learning_rate': 0.05784542294445882, 'subsample': 0.6105451402652029, 'colsample_bytree': 0.7391313957038773, 'gamma': 1.7467184839854677, 'scale_pos_weight': 3.542755215442128}. Best is trial 22 with value: 0.660352807351118.


🏃 View run judicious-hare-614 at: http://127.0.0.1:5000/#/experiments/3/runs/1635867c895d4d21b357c19aeb829077
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:02,246] Trial 31 finished with value: 0.6578994476598563 and parameters: {'n_estimators': 814, 'max_depth': 5, 'learning_rate': 0.057064912166962176, 'subsample': 0.7789329685343261, 'colsample_bytree': 0.598105075161297, 'gamma': 4.7089340513641, 'scale_pos_weight': 3.068460556149384}. Best is trial 22 with value: 0.660352807351118.


🏃 View run trusting-cub-499 at: http://127.0.0.1:5000/#/experiments/3/runs/913fd2a0e75f4c06b08ca0aac78848ef
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:03,621] Trial 32 finished with value: 0.6579962582346963 and parameters: {'n_estimators': 823, 'max_depth': 5, 'learning_rate': 0.03921026477441133, 'subsample': 0.8329666429900773, 'colsample_bytree': 0.5475015309584584, 'gamma': 4.7926004937884885, 'scale_pos_weight': 3.2807518875166073}. Best is trial 22 with value: 0.660352807351118.


🏃 View run ambitious-shrew-883 at: http://127.0.0.1:5000/#/experiments/3/runs/69025234198f4d328b07b24662d89ae8
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:05,117] Trial 33 finished with value: 0.6592750287029895 and parameters: {'n_estimators': 927, 'max_depth': 6, 'learning_rate': 0.052957970247760486, 'subsample': 0.7355799409323512, 'colsample_bytree': 0.6982555206885416, 'gamma': 4.298581805571373, 'scale_pos_weight': 1.3512646657799325}. Best is trial 22 with value: 0.660352807351118.


🏃 View run efficient-stag-679 at: http://127.0.0.1:5000/#/experiments/3/runs/c46e1c83da8b4579869d1331d787c61f
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:06,618] Trial 34 finished with value: 0.6599666720515619 and parameters: {'n_estimators': 932, 'max_depth': 9, 'learning_rate': 0.08145676993670734, 'subsample': 0.6757107423732369, 'colsample_bytree': 0.7034291048451576, 'gamma': 4.29561152328904, 'scale_pos_weight': 1.3134460331893365}. Best is trial 22 with value: 0.660352807351118.


🏃 View run bright-midge-905 at: http://127.0.0.1:5000/#/experiments/3/runs/aaeeee12169a444e8ea7203b98441c9f
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:08,995] Trial 35 finished with value: 0.6629579698712056 and parameters: {'n_estimators': 930, 'max_depth': 9, 'learning_rate': 0.015319037940069946, 'subsample': 0.5749650681035079, 'colsample_bytree': 0.7077627844451623, 'gamma': 4.350438030915925, 'scale_pos_weight': 1.2756150958086228}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run luminous-steed-268 at: http://127.0.0.1:5000/#/experiments/3/runs/b5abfdf26d534fa3827c07fe8433c212
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:10,710] Trial 36 finished with value: 0.6594309785173496 and parameters: {'n_estimators': 982, 'max_depth': 9, 'learning_rate': 0.015038568753006054, 'subsample': 0.5645541584301791, 'colsample_bytree': 0.8331948006013692, 'gamma': 3.654071784671834, 'scale_pos_weight': 1.2343996253054237}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run nebulous-horse-892 at: http://127.0.0.1:5000/#/experiments/3/runs/342d1a3f5cfe42f7b4303a9c83f19e5c
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:12,379] Trial 37 finished with value: 0.6596462203426539 and parameters: {'n_estimators': 944, 'max_depth': 10, 'learning_rate': 0.01036531794118666, 'subsample': 0.6782783900625597, 'colsample_bytree': 0.7894046958092614, 'gamma': 4.484614498630832, 'scale_pos_weight': 1.5350704696822088}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run unleashed-quail-868 at: http://127.0.0.1:5000/#/experiments/3/runs/e053a0f567034a5289b94539df549b34
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:13,888] Trial 38 finished with value: 0.6618097475725049 and parameters: {'n_estimators': 880, 'max_depth': 9, 'learning_rate': 0.01583320251967311, 'subsample': 0.5298165998899604, 'colsample_bytree': 0.6777234152179529, 'gamma': 4.2257612740550785, 'scale_pos_weight': 1.149831736526099}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run placid-shrew-329 at: http://127.0.0.1:5000/#/experiments/3/runs/0e54019ee3cb4a0bbff95e63bce2ffa2
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:15,498] Trial 39 finished with value: 0.6615332386961112 and parameters: {'n_estimators': 854, 'max_depth': 9, 'learning_rate': 0.01633681746356878, 'subsample': 0.5592922323466736, 'colsample_bytree': 0.6726578348337051, 'gamma': 4.172282880606545, 'scale_pos_weight': 1.228176358619098}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run auspicious-mare-68 at: http://127.0.0.1:5000/#/experiments/3/runs/62fbd807553c424c9d335c82fcbdeb5e
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:17,268] Trial 40 finished with value: 0.6623847395449117 and parameters: {'n_estimators': 874, 'max_depth': 11, 'learning_rate': 0.01610135389398656, 'subsample': 0.5537208644269445, 'colsample_bytree': 0.6728077494008259, 'gamma': 3.733519039842342, 'scale_pos_weight': 1.039551441366607}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run resilient-moth-482 at: http://127.0.0.1:5000/#/experiments/3/runs/6fdfad21cac1486094c7b67138c19bff
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:18,914] Trial 41 finished with value: 0.662927566173188 and parameters: {'n_estimators': 877, 'max_depth': 11, 'learning_rate': 0.0164599824858276, 'subsample': 0.5572189772403493, 'colsample_bytree': 0.6729910082895892, 'gamma': 3.7646939194632685, 'scale_pos_weight': 1.0156393353565316}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run respected-croc-892 at: http://127.0.0.1:5000/#/experiments/3/runs/03748756125a4d43889ce4e4297d43f8
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:20,600] Trial 42 finished with value: 0.662703525716203 and parameters: {'n_estimators': 854, 'max_depth': 11, 'learning_rate': 0.01608122473813707, 'subsample': 0.5547307257427039, 'colsample_bytree': 0.6827291111722616, 'gamma': 3.7381236777569273, 'scale_pos_weight': 1.034935150065925}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run spiffy-dog-153 at: http://127.0.0.1:5000/#/experiments/3/runs/5473ef1e265f4bef8b21dafd27149937
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:22,199] Trial 43 finished with value: 0.6616071973222893 and parameters: {'n_estimators': 728, 'max_depth': 12, 'learning_rate': 0.01320537985825916, 'subsample': 0.500774090311553, 'colsample_bytree': 0.7327222300332555, 'gamma': 3.6583888836796414, 'scale_pos_weight': 1.0249622728440602}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run rebellious-wolf-635 at: http://127.0.0.1:5000/#/experiments/3/runs/3f20183f2c7542d7a3d098f01d92f7f7
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:24,521] Trial 44 finished with value: 0.6593181601468632 and parameters: {'n_estimators': 1000, 'max_depth': 11, 'learning_rate': 0.018555800895390342, 'subsample': 0.5374231732377567, 'colsample_bytree': 0.676639546883505, 'gamma': 3.0161679617670982, 'scale_pos_weight': 1.1360896623493353}. Best is trial 35 with value: 0.6629579698712056.


🏃 View run fun-cod-565 at: http://127.0.0.1:5000/#/experiments/3/runs/e5aca1c61f6b4629838a636d13856651
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:26,204] Trial 45 finished with value: 0.6634445228686047 and parameters: {'n_estimators': 639, 'max_depth': 12, 'learning_rate': 0.011215038192750089, 'subsample': 0.6256223590894531, 'colsample_bytree': 0.7243465457053382, 'gamma': 3.1826154221071326, 'scale_pos_weight': 1.0132970924196423}. Best is trial 45 with value: 0.6634445228686047.


🏃 View run invincible-fox-572 at: http://127.0.0.1:5000/#/experiments/3/runs/ecb625fd66a94a1db73e9e82f0a4baf7
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:27,845] Trial 46 finished with value: 0.6607447308267345 and parameters: {'n_estimators': 628, 'max_depth': 12, 'learning_rate': 0.010839753692654995, 'subsample': 0.6288413492323808, 'colsample_bytree': 0.7669255156012658, 'gamma': 3.2449704800176096, 'scale_pos_weight': 1.0213206744927785}. Best is trial 45 with value: 0.6634445228686047.


🏃 View run nervous-auk-602 at: http://127.0.0.1:5000/#/experiments/3/runs/9c9f606c5f214fd99db3d42bf19a3553
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:29,284] Trial 47 finished with value: 0.6615233334734416 and parameters: {'n_estimators': 516, 'max_depth': 12, 'learning_rate': 0.012667228504241596, 'subsample': 0.5898031919624189, 'colsample_bytree': 0.7215101061728911, 'gamma': 3.5033071921238608, 'scale_pos_weight': 1.366267849398435}. Best is trial 45 with value: 0.6634445228686047.


🏃 View run peaceful-frog-850 at: http://127.0.0.1:5000/#/experiments/3/runs/d7832f70c8544feb93acf58abb18b0b7
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:30,800] Trial 48 finished with value: 0.6593069074782336 and parameters: {'n_estimators': 730, 'max_depth': 11, 'learning_rate': 0.027370387609323664, 'subsample': 0.5772595051070301, 'colsample_bytree': 0.7611943787366936, 'gamma': 3.2095034511515084, 'scale_pos_weight': 1.0106987054631333}. Best is trial 45 with value: 0.6634445228686047.


🏃 View run adventurous-bat-611 at: http://127.0.0.1:5000/#/experiments/3/runs/8c6bc5dae11a45b7ba8e7dff749624f4
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:50:31,895] Trial 49 finished with value: 0.6566332748976177 and parameters: {'n_estimators': 351, 'max_depth': 13, 'learning_rate': 0.024869990578416996, 'subsample': 0.5382957271699005, 'colsample_bytree': 0.6525915460932475, 'gamma': 3.8016264314149173, 'scale_pos_weight': 1.5949132412249751}. Best is trial 45 with value: 0.6634445228686047.


🏃 View run dapper-goat-564 at: http://127.0.0.1:5000/#/experiments/3/runs/87bcd0a9cd394ce8800df8fe626fa08a
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3
----------------------------------------
XGBoost tuning completed!
Best CV PR-AUC: 0.6634
Validation PR-AUC: 0.6810
Best Parameters: {'n_estimators': 639, 'max_depth': 12, 'learning_rate': 0.011215038192750089, 'subsample': 0.6256223590894531, 'colsample_bytree': 0.7243465457053382, 'gamma': 3.1826154221071326, 'scale_pos_weight': 1.0132970924196423}
🏃 View run XGBoost_PR_AUC_Optimization_CV at: http://127.0.0.1:5000/#/experiments/3/runs/ed0546e8b4ee4d13ba29355065792f4a
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


In [ ]:
# # Vẽ ma trận nhầm lẫn chuẩn hóa (normalized) theo hàng ngang (true labels)
# fig, ax = plt.subplots(figsize=(8, 6))
# disp_norm = ConfusionMatrixDisplay.from_estimator(
#     best_xgb,
#     X_test,
#     y_test,
#     display_labels=["Active (0)", "Churn (1)"],
#     cmap="Greens",
#     normalize='true', # Quan trọng: tính % theo từng lớp thực tế
#     ax=ax
# )

# plt.title("Normalized Confusion Matrix (%)")
# plt.show()

### Catboost

In [52]:
default_catboost = CatBoostClassifier(
    random_state=RANDOM_STATE,
    loss_function="Logloss",
    eval_metric="PRAUC",
    auto_class_weights="Balanced",
    verbose=0,
    allow_writing_files=False,
)
default_catboost.fit(X_train_tree, y_train_tree) 
y_val_proba = default_catboost.predict_proba(X_val_tree)[:, 1] 
y_val_pred = (y_val_proba >= 0.5).astype(int) 
print("CatBoost Default") 
print("PR-AUC:", round(average_precision_score(y_val_tree, y_val_proba), 4)) 
print("ROC-AUC:", round(roc_auc_score(y_val_tree, y_val_proba), 4)) 
print("LogLoss:", round(log_loss(y_val_tree, y_val_proba), 4)) 
print("Accuracy:", round(accuracy_score(y_val_tree, y_val_pred), 4)) 
print("Precision:", round(precision_score(y_val_tree, y_val_pred, zero_division=0), 4)) 
print("Recall:", round(recall_score(y_val_tree, y_val_pred, zero_division=0), 4)) 
print("F1:", round(f1_score(y_val_tree, y_val_pred, zero_division=0), 4)) 
print("PPR:", round(y_val_pred.mean(), 4))

CatBoost Default
PR-AUC: 0.6617
ROC-AUC: 0.8402
LogLoss: 0.4653
Accuracy: 0.7817
Precision: 0.5678
Recall: 0.7425
F1: 0.6435
PPR: 0.3469


In [33]:
# =========================================
# CatBoost tuning with CV
# =========================================
def objective_catboost(trial):
    """
    Tune CatBoost using cross-validated PR-AUC on the training set only.
    This avoids validation/test leakage during hyperparameter search.
    """
    params = {
        "iterations": trial.suggest_int("iterations", 200, 1000),
        "depth": trial.suggest_int("depth", 4, 10),
        "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.2, log=True),
        "l2_leaf_reg": trial.suggest_float("l2_leaf_reg", 1.0, 10.0),
        "subsample": trial.suggest_float("subsample", 0.5, 1.0),
        "random_strength": trial.suggest_float("random_strength", 0.0, 5.0),
        "scale_pos_weight": trial.suggest_float("scale_pos_weight", 1.0, ratio_tree * 1.5),
        "random_state": RANDOM_STATE,
        "loss_function": "Logloss",
        "eval_metric": "PRAUC",
        "verbose": 0,
        "allow_writing_files": False,
    }

    model = CatBoostClassifier(**params)

    y_proba_oof = cross_val_predict(
        estimator=model,
        X=X_train_tree,
        y=y_train_tree,
        cv=cv_tree,
        method="predict_proba",
    )[:, 1]

    y_pred_oof = (y_proba_oof >= 0.5).astype(int)

    metrics = {
        "cv_pr_auc": float(average_precision_score(y_train_tree, y_proba_oof)),
        "cv_roc_auc": float(roc_auc_score(y_train_tree, y_proba_oof)),
        "cv_log_loss": float(log_loss(y_train_tree, y_proba_oof)),
        "cv_accuracy": float(accuracy_score(y_train_tree, y_pred_oof)),
        "cv_precision": float(precision_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_recall": float(recall_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_f1_score": float(f1_score(y_train_tree, y_pred_oof, zero_division=0)),
        "cv_predicted_positive_rate": float(y_pred_oof.mean()),
    }

    with mlflow.start_run(nested=True):
        mlflow.log_params(params)
        mlflow.log_metrics(metrics)

    return metrics["cv_pr_auc"]


with mlflow.start_run(run_name="CatBoost_PR_AUC_Optimization_CV"):
    logger.info("Starting CatBoost hyperparameter tuning with Stratified CV...")

    study_catboost = optuna.create_study(direction="maximize")
    study_catboost.optimize(objective_catboost, n_trials=30)

    mlflow.log_params(study_catboost.best_params)
    mlflow.log_metric("best_cv_pr_auc", float(study_catboost.best_value))
    mlflow.log_params({
        "dataset_type": "tree",
        "default_threshold": 0.5,
    })

    best_catboost = CatBoostClassifier(
        **study_catboost.best_params,
        random_state=RANDOM_STATE,
        loss_function="Logloss",
        eval_metric="PRAUC",
        verbose=0,
        allow_writing_files=False,
    )
    best_catboost.fit(X_train_tree, y_train_tree)

    # Evaluate on validation set only at this stage
    y_val_proba = best_catboost.predict_proba(X_val_tree)[:, 1]
    y_val_pred = (y_val_proba >= 0.5).astype(int)

    val_metrics = {
        "val_pr_auc": float(average_precision_score(y_val_tree, y_val_proba)),
        "val_roc_auc": float(roc_auc_score(y_val_tree, y_val_proba)),
        "val_log_loss": float(log_loss(y_val_tree, y_val_proba)),
        "val_accuracy": float(accuracy_score(y_val_tree, y_val_pred)),
        "val_precision": float(precision_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_recall": float(recall_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_f1_score": float(f1_score(y_val_tree, y_val_pred, zero_division=0)),
        "val_predicted_positive_rate": float(y_val_pred.mean()),
    }

    mlflow.log_metrics(val_metrics)
    mlflow.sklearn.log_model(best_catboost, name="best_catboost_prauc_model")

    overall_results["CatBoost"] = {
        "model": best_catboost,
        "dataset_type": "tree",
        "cv_pr_auc": float(study_catboost.best_value),
        "val_pr_auc": float(val_metrics["val_pr_auc"]),
        "val_roc_auc": float(val_metrics["val_roc_auc"]),
        "val_log_loss": float(val_metrics["val_log_loss"]),
        "accuracy": float(val_metrics["val_accuracy"]),
        "precision": float(val_metrics["val_precision"]),
        "recall": float(val_metrics["val_recall"]),
        "f1_score": float(val_metrics["val_f1_score"]),
        "predicted_positive_rate": float(val_metrics["val_predicted_positive_rate"]),
        "threshold": 0.5,
        "params": study_catboost.best_params,
    }

    print("-" * 40)
    print("CatBoost tuning completed!")
    print(f"Best CV PR-AUC: {study_catboost.best_value:.4f}")
    print(f"Validation PR-AUC: {val_metrics['val_pr_auc']:.4f}")
    print(f"Best Parameters: {study_catboost.best_params}")

2026-04-15 15:51:46 [INFO] Starting CatBoost hyperparameter tuning with Stratified CV...
[I 2026-04-15 15:51:46,968] A new study created in memory with name: no-name-78d6bd53-ade9-480a-9dab-2006e2cdf723
[I 2026-04-15 15:52:01,736] Trial 0 finished with value: 0.6518696162870378 and parameters: {'iterations': 646, 'depth': 5, 'learning_rate': 0.03815550183215712, 'l2_leaf_reg': 6.306189157421232, 'subsample': 0.8070792563506604, 'random_strength': 4.604615079362939, 'scale_pos_weight': 3.82019419854957}. Best is trial 0 with value: 0.6518696162870378.


🏃 View run tasteful-seal-784 at: http://127.0.0.1:5000/#/experiments/3/runs/66d89a4bfc2c42fca1e2e07be5da33c3
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:53:19,169] Trial 1 finished with value: 0.6242378611336606 and parameters: {'iterations': 710, 'depth': 10, 'learning_rate': 0.04256774975394579, 'l2_leaf_reg': 8.574773526016305, 'subsample': 0.8587020333555975, 'random_strength': 4.722208146202751, 'scale_pos_weight': 2.3819043179051222}. Best is trial 0 with value: 0.6518696162870378.


🏃 View run beautiful-wren-153 at: http://127.0.0.1:5000/#/experiments/3/runs/8d393ef1507e4a918a71a2b84fb36418
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:53:52,335] Trial 2 finished with value: 0.6284102680021022 and parameters: {'iterations': 988, 'depth': 7, 'learning_rate': 0.04913307022014408, 'l2_leaf_reg': 5.6549295460973195, 'subsample': 0.7132487315978067, 'random_strength': 2.673640458665967, 'scale_pos_weight': 1.125404769972028}. Best is trial 0 with value: 0.6518696162870378.


🏃 View run popular-lark-612 at: http://127.0.0.1:5000/#/experiments/3/runs/c1c60993030b40f5ab51a062670602c0
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:54:18,077] Trial 3 finished with value: 0.6104921710698631 and parameters: {'iterations': 993, 'depth': 6, 'learning_rate': 0.12423858711578564, 'l2_leaf_reg': 5.893373648156889, 'subsample': 0.8285983849466667, 'random_strength': 1.7844136682663354, 'scale_pos_weight': 2.9847777190091}. Best is trial 0 with value: 0.6518696162870378.


🏃 View run unleashed-frog-435 at: http://127.0.0.1:5000/#/experiments/3/runs/a5187af5a1064591a7cdd0c4bee2bec7
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:54:39,512] Trial 4 finished with value: 0.6368683145512533 and parameters: {'iterations': 484, 'depth': 8, 'learning_rate': 0.05894018750359419, 'l2_leaf_reg': 4.8151796191942235, 'subsample': 0.7062601352840643, 'random_strength': 3.075437849765045, 'scale_pos_weight': 3.42788863485751}. Best is trial 0 with value: 0.6518696162870378.


🏃 View run skillful-trout-191 at: http://127.0.0.1:5000/#/experiments/3/runs/797a2bfa24304178b2bbb3df5840a5c5
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:56:04,645] Trial 5 finished with value: 0.6051398191800417 and parameters: {'iterations': 785, 'depth': 10, 'learning_rate': 0.18867128327496377, 'l2_leaf_reg': 9.471034848121702, 'subsample': 0.6559569280113313, 'random_strength': 4.143775672508315, 'scale_pos_weight': 2.4189932742060556}. Best is trial 0 with value: 0.6518696162870378.


🏃 View run persistent-whale-789 at: http://127.0.0.1:5000/#/experiments/3/runs/6cce8325b48c4b3ea56f760588dae6eb
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:56:21,708] Trial 6 finished with value: 0.6619441213922026 and parameters: {'iterations': 690, 'depth': 6, 'learning_rate': 0.013652985030740214, 'l2_leaf_reg': 3.90768121881874, 'subsample': 0.8293583404558196, 'random_strength': 4.769372672161012, 'scale_pos_weight': 3.3898794175721303}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run serious-boar-816 at: http://127.0.0.1:5000/#/experiments/3/runs/8bf00e00cc6a4b1b82dc966db4d70544
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:57:47,944] Trial 7 finished with value: 0.6437612639367106 and parameters: {'iterations': 882, 'depth': 10, 'learning_rate': 0.01025118013351109, 'l2_leaf_reg': 8.382203947959747, 'subsample': 0.6611731151394518, 'random_strength': 1.589312930947877, 'scale_pos_weight': 2.469270256310802}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run melodic-gnat-810 at: http://127.0.0.1:5000/#/experiments/3/runs/9d1e338ea1414a3887b2dab2c364cb85
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:58:57,019] Trial 8 finished with value: 0.601810365392408 and parameters: {'iterations': 622, 'depth': 10, 'learning_rate': 0.18706390943486348, 'l2_leaf_reg': 5.996430158336548, 'subsample': 0.5047187669951929, 'random_strength': 3.756375432092077, 'scale_pos_weight': 3.9840891674067636}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run legendary-midge-282 at: http://127.0.0.1:5000/#/experiments/3/runs/9b74b3879d544642818477c17dddee7a
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:59:19,031] Trial 9 finished with value: 0.6332405347364268 and parameters: {'iterations': 322, 'depth': 9, 'learning_rate': 0.06689842104218607, 'l2_leaf_reg': 2.2913928407870587, 'subsample': 0.600289259572748, 'random_strength': 2.376352179028192, 'scale_pos_weight': 3.9607912804770953}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run enchanting-moth-935 at: http://127.0.0.1:5000/#/experiments/3/runs/f8f0c69a039841b09ccf8d169218685d
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:59:27,426] Trial 10 finished with value: 0.6592239005268703 and parameters: {'iterations': 443, 'depth': 4, 'learning_rate': 0.01091031760929173, 'l2_leaf_reg': 1.015939190624457, 'subsample': 0.9833952154210437, 'random_strength': 0.10956472514537197, 'scale_pos_weight': 1.565962045140052}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run rumbling-crab-476 at: http://127.0.0.1:5000/#/experiments/3/runs/8947895162264f71b28d7ec13523182f
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:59:35,809] Trial 11 finished with value: 0.6608489298146333 and parameters: {'iterations': 443, 'depth': 4, 'learning_rate': 0.010249286921624363, 'l2_leaf_reg': 1.2113046688059512, 'subsample': 0.9587346576690126, 'random_strength': 0.37844309181010866, 'scale_pos_weight': 1.3998022965002526}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run dazzling-calf-514 at: http://127.0.0.1:5000/#/experiments/3/runs/bf54d62c7b2a4c8eb61a76fa288ffd9e
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:59:40,566] Trial 12 finished with value: 0.6593268857000261 and parameters: {'iterations': 230, 'depth': 4, 'learning_rate': 0.01862852933253981, 'l2_leaf_reg': 3.3766365639485723, 'subsample': 0.9669534701507939, 'random_strength': 0.19374280730056048, 'scale_pos_weight': 1.5749453935297806}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run delicate-rat-433 at: http://127.0.0.1:5000/#/experiments/3/runs/90c445d8ef744d3dbd74a4d24b03fa09
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 15:59:52,933] Trial 13 finished with value: 0.6605186938766211 and parameters: {'iterations': 476, 'depth': 6, 'learning_rate': 0.019557022220809867, 'l2_leaf_reg': 3.5391844696924912, 'subsample': 0.906957496333649, 'random_strength': 1.0032517979480056, 'scale_pos_weight': 3.0076367664120456}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run adaptable-rat-140 at: http://127.0.0.1:5000/#/experiments/3/runs/54efed9bae6043aa8cd78b0ef6989799
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:00:04,910] Trial 14 finished with value: 0.6605052666616303 and parameters: {'iterations': 548, 'depth': 5, 'learning_rate': 0.018126590845761584, 'l2_leaf_reg': 1.2722958240954656, 'subsample': 0.8724124656599563, 'random_strength': 3.3663285780183294, 'scale_pos_weight': 1.6205514530579905}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run spiffy-cat-409 at: http://127.0.0.1:5000/#/experiments/3/runs/507278a3eace4e4e9c4a1185feba18ad
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:00:14,519] Trial 15 finished with value: 0.6573017079282366 and parameters: {'iterations': 364, 'depth': 6, 'learning_rate': 0.028485365329338274, 'l2_leaf_reg': 3.5138917860588212, 'subsample': 0.9269837001266472, 'random_strength': 0.8095651615346702, 'scale_pos_weight': 2.0110771657511024}. Best is trial 6 with value: 0.6619441213922026.


🏃 View run placid-hen-217 at: http://127.0.0.1:5000/#/experiments/3/runs/8587d489812f4619bf92b22f8f4fc4cc
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:00:31,602] Trial 16 finished with value: 0.6638036206796916 and parameters: {'iterations': 784, 'depth': 5, 'learning_rate': 0.014347042575565043, 'l2_leaf_reg': 2.27104850519003, 'subsample': 0.7783425489210776, 'random_strength': 2.173911534359554, 'scale_pos_weight': 3.1764274646112436}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run angry-shrimp-54 at: http://127.0.0.1:5000/#/experiments/3/runs/d8773c4881804316bbe7ada0654a03d1
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:00:56,502] Trial 17 finished with value: 0.6473199493715908 and parameters: {'iterations': 784, 'depth': 7, 'learning_rate': 0.025653743262188698, 'l2_leaf_reg': 2.613234061922756, 'subsample': 0.784788358081198, 'random_strength': 2.0193030860346, 'scale_pos_weight': 3.357367609450925}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run wistful-cub-573 at: http://127.0.0.1:5000/#/experiments/3/runs/814ab4e4e703489f9fa599df23573b67
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:01:15,681] Trial 18 finished with value: 0.6621346474704579 and parameters: {'iterations': 877, 'depth': 5, 'learning_rate': 0.013508975820663013, 'l2_leaf_reg': 4.859342465018814, 'subsample': 0.7577636783710789, 'random_strength': 2.72604713119568, 'scale_pos_weight': 2.8948307306161776}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run able-penguin-853 at: http://127.0.0.1:5000/#/experiments/3/runs/f09433ad443b43e2a85a74494acb3ce6
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:01:34,827] Trial 19 finished with value: 0.6528224875321547 and parameters: {'iterations': 871, 'depth': 5, 'learning_rate': 0.026896703730852126, 'l2_leaf_reg': 7.152928512854739, 'subsample': 0.7589878947261893, 'random_strength': 2.8563198616829615, 'scale_pos_weight': 2.884602818626107}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run mysterious-skunk-557 at: http://127.0.0.1:5000/#/experiments/3/runs/79b4d50b53d84cb39f30b99efa188ed4
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:02:12,416] Trial 20 finished with value: 0.6468339422326295 and parameters: {'iterations': 880, 'depth': 8, 'learning_rate': 0.014586139144982758, 'l2_leaf_reg': 4.641982665251738, 'subsample': 0.5774008125881469, 'random_strength': 1.2356140074634747, 'scale_pos_weight': 2.1314501153830245}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run auspicious-cub-62 at: http://127.0.0.1:5000/#/experiments/3/runs/b814dd75150f4c468100a65b9afcbe18
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:02:28,991] Trial 21 finished with value: 0.6626440274594436 and parameters: {'iterations': 745, 'depth': 5, 'learning_rate': 0.01424581714356726, 'l2_leaf_reg': 4.597337447637438, 'subsample': 0.7287098246171043, 'random_strength': 3.8532523245694232, 'scale_pos_weight': 3.4427765602825757}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run bouncy-stoat-351 at: http://127.0.0.1:5000/#/experiments/3/runs/1f68872885d34e8ebe3fff80ecb6120a
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:02:47,880] Trial 22 finished with value: 0.6628176380079172 and parameters: {'iterations': 791, 'depth': 5, 'learning_rate': 0.014733590894447422, 'l2_leaf_reg': 4.673410147617351, 'subsample': 0.7275061144516983, 'random_strength': 3.4683878875534067, 'scale_pos_weight': 3.6645247469494207}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run carefree-whale-880 at: http://127.0.0.1:5000/#/experiments/3/runs/82f48b73b81740d89b01ffad185b1704
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:03:09,054] Trial 23 finished with value: 0.6589993069548598 and parameters: {'iterations': 750, 'depth': 5, 'learning_rate': 0.021190538518310018, 'l2_leaf_reg': 2.5898838321731334, 'subsample': 0.720993024002383, 'random_strength': 3.827105269864473, 'scale_pos_weight': 3.669421941322355}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run clean-frog-726 at: http://127.0.0.1:5000/#/experiments/3/runs/5ffed2a7ca6444379c860806efcc7073
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:03:27,217] Trial 24 finished with value: 0.6634058467993614 and parameters: {'iterations': 812, 'depth': 4, 'learning_rate': 0.015330398658446997, 'l2_leaf_reg': 6.647519529405022, 'subsample': 0.681702490441838, 'random_strength': 3.314120175337179, 'scale_pos_weight': 3.6086487683874635}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run gregarious-koi-507 at: http://127.0.0.1:5000/#/experiments/3/runs/a096f72d70764f13a84d4811ec105b2f
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:03:44,028] Trial 25 finished with value: 0.6391439532241887 and parameters: {'iterations': 827, 'depth': 4, 'learning_rate': 0.08340868853067673, 'l2_leaf_reg': 7.2024610540618825, 'subsample': 0.6651683733896694, 'random_strength': 2.25017232142032, 'scale_pos_weight': 3.650505476604714}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run shivering-hawk-728 at: http://127.0.0.1:5000/#/experiments/3/runs/e3f95171b3db42f39532e81dfd895357
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:04:02,645] Trial 26 finished with value: 0.6541049610834796 and parameters: {'iterations': 932, 'depth': 4, 'learning_rate': 0.033408354444337644, 'l2_leaf_reg': 6.9719432468039155, 'subsample': 0.59249652383622, 'random_strength': 3.3221591395092713, 'scale_pos_weight': 3.181775806346377}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run popular-loon-776 at: http://127.0.0.1:5000/#/experiments/3/runs/e82d26137bce415aa2e7741d48587a9d
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:04:18,392] Trial 27 finished with value: 0.6585743341803394 and parameters: {'iterations': 572, 'depth': 6, 'learning_rate': 0.023006197627653335, 'l2_leaf_reg': 1.9003206284543406, 'subsample': 0.6864291995217533, 'random_strength': 3.192056157909587, 'scale_pos_weight': 3.693601841170978}. Best is trial 16 with value: 0.6638036206796916.


🏃 View run wistful-rook-763 at: http://127.0.0.1:5000/#/experiments/3/runs/05df844c48ca42c1a70ca40445b29d4e
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:04:31,517] Trial 28 finished with value: 0.6638495848075624 and parameters: {'iterations': 666, 'depth': 4, 'learning_rate': 0.016689585012689682, 'l2_leaf_reg': 8.06524410655627, 'subsample': 0.6361550152229549, 'random_strength': 3.575977810483283, 'scale_pos_weight': 2.7446507593224845}. Best is trial 28 with value: 0.6638495848075624.


🏃 View run delightful-shad-146 at: http://127.0.0.1:5000/#/experiments/3/runs/9e5a6f199bd14630b01bf01f4433810b
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


[I 2026-04-15 16:04:44,871] Trial 29 finished with value: 0.6588135841591616 and parameters: {'iterations': 698, 'depth': 4, 'learning_rate': 0.03476091620740038, 'l2_leaf_reg': 8.034301881857527, 'subsample': 0.5451181008309505, 'random_strength': 4.312379190183973, 'scale_pos_weight': 2.6961865619975542}. Best is trial 28 with value: 0.6638495848075624.


🏃 View run unleashed-sloth-30 at: http://127.0.0.1:5000/#/experiments/3/runs/94b029a4848b46e1ab8e85adbd6f13d9
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


2026/04/15 16:04:48 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


----------------------------------------
CatBoost tuning completed!
Best CV PR-AUC: 0.6638
Validation PR-AUC: 0.6858
Best Parameters: {'iterations': 666, 'depth': 4, 'learning_rate': 0.016689585012689682, 'l2_leaf_reg': 8.06524410655627, 'subsample': 0.6361550152229549, 'random_strength': 3.575977810483283, 'scale_pos_weight': 2.7446507593224845}
🏃 View run CatBoost_PR_AUC_Optimization_CV at: http://127.0.0.1:5000/#/experiments/3/runs/ff6d0149ff9c4445a9ccc7d1f4bafe2f
🧪 View experiment at: http://127.0.0.1:5000/#/experiments/3


In [34]:
#### Backfill additional evaluation metrics for trained models

DEFAULT_THRESHOLD = 0.5

for model_name, info in overall_results.items():
    model = info["model"]
    dataset_type = info["dataset_type"]

    if dataset_type == "linear":
        X_eval = X_test_linear
        y_eval = y_test_linear
    elif dataset_type == "tree":
        X_eval = X_test_tree
        y_eval = y_test_tree
    else:
        raise ValueError(f"Unknown dataset_type for model {model_name}: {dataset_type}")

    y_test_proba = model.predict_proba(X_eval)[:, 1]
    y_test_pred = (y_test_proba >= DEFAULT_THRESHOLD).astype(int)

    info.update({
        "test_pr_auc": float(average_precision_score(y_eval, y_test_proba)),
        "accuracy": float(accuracy_score(y_eval, y_test_pred)),
        "precision": float(precision_score(y_eval, y_test_pred, zero_division=0)),
        "recall": float(recall_score(y_eval, y_test_pred, zero_division=0)),
        "f1_score": float(f1_score(y_eval, y_test_pred, zero_division=0)),
        "roc_auc": float(roc_auc_score(y_eval, y_test_proba)),
        "log_loss": float(log_loss(y_eval, y_test_proba)),
        "predicted_positive_rate": float(y_test_pred.mean()),
        "threshold": float(DEFAULT_THRESHOLD),
    })

print("Updated overall_results with additional test metrics.")
for model_name, info in overall_results.items():
    print(
        model_name,
        "| Dataset:", info["dataset_type"],
        "| PR-AUC:", round(info["test_pr_auc"], 4),
        "| Acc:", round(info["accuracy"], 4),
        "| Prec:", round(info["precision"], 4),
        "| Rec:", round(info["recall"], 4),
        "| F1:", round(info["f1_score"], 4),
        "| ROC-AUC:", round(info["roc_auc"], 4),
        "| LogLoss:", round(info["log_loss"], 4),
        "| PPR:", round(info["predicted_positive_rate"], 4),
    )

Updated overall_results with additional test metrics.
Baseline_LogisticRegression | Dataset: linear | PR-AUC: 0.6349 | Acc: 0.7445 | Prec: 0.5123 | Rec: 0.7807 | F1: 0.6186 | ROC-AUC: 0.8396 | LogLoss: 0.4968 | PPR: 0.4045
RandomForest | Dataset: tree | PR-AUC: 0.6446 | Acc: 0.775 | Prec: 0.5573 | Rec: 0.7406 | F1: 0.6361 | ROC-AUC: 0.8392 | LogLoss: 0.4692 | PPR: 0.3527
XGBoost | Dataset: tree | PR-AUC: 0.6474 | Acc: 0.8041 | Prec: 0.6561 | Rec: 0.5508 | F1: 0.5988 | ROC-AUC: 0.8431 | LogLoss: 0.4193 | PPR: 0.2229
CatBoost | Dataset: tree | PR-AUC: 0.6584 | Acc: 0.7537 | Prec: 0.5236 | Rec: 0.8021 | F1: 0.6336 | ROC-AUC: 0.8444 | LogLoss: 0.4842 | PPR: 0.4067


## Candidate model selection for threshold tuning

CatBoost achieves the highest PR-AUC (0.6584) and the highest recall (0.8021), indicating strong ability to identify churn-risk customers. However, it has relatively lower precision and a higher predicted positive rate, suggesting a more aggressive targeting strategy.

On the other hand, XGBoost achieves slightly lower PR-AUC (0.6474) but significantly higher precision (0.6561) and a much lower predicted positive rate. This indicates a more conservative approach, focusing on fewer but more confident churn predictions.

Since these two models represent different trade-offs between recall and precision, both are retained for threshold tuning. This allows selecting the final model-threshold combination based on business-oriented considerations rather than ranking performance alone.

## Threshold tuning

While PR-AUC evaluates the ranking quality of predicted probabilities, the final classification outcome depends on the decision threshold.

In the context of telco churn prediction:
- Lower thresholds increase recall, allowing the model to capture more potential churn cases.
- Higher thresholds increase precision, reducing unnecessary retention actions.

Therefore, threshold tuning is performed on the validation set to identify the optimal trade-off between recall, precision, and predicted positive rate.

The following analysis evaluates CatBoost and XGBoost across a range of thresholds.

In [20]:
thresholds = np.round(np.arange(0.10, 0.91, 0.05), 2)
thresholds

array([0.1 , 0.15, 0.2 , 0.25, 0.3 , 0.35, 0.4 , 0.45, 0.5 , 0.55, 0.6 ,
       0.65, 0.7 , 0.75, 0.8 , 0.85, 0.9 ])

In [38]:
def evaluate_thresholds(model, X, y, thresholds):
    y_proba = model.predict_proba(X)[:, 1]
    rows = []

    for t in thresholds:
        y_pred = (y_proba >= t).astype(int)

        rows.append({
            "threshold": t,
            "accuracy": accuracy_score(y, y_pred),
            "precision": precision_score(y, y_pred, zero_division=0),
            "recall": recall_score(y, y_pred, zero_division=0),
            "f1_score": f1_score(y, y_pred, zero_division=0),
            "predicted_positive_rate": y_pred.mean(),
        })

    return pd.DataFrame(rows)

In [40]:
threshold_results = {}

for model_name in ["CatBoost", "XGBoost"]:
    info = overall_results[model_name]
    model = info["model"]

    threshold_df = evaluate_thresholds(model, X_val_tree, y_val_tree, thresholds)
    threshold_df.insert(0, "model", model_name)

    threshold_results[model_name] = threshold_df

In [41]:
print("CatBoost")
display(threshold_results["CatBoost"])

print("XGBoost")
display(threshold_results["XGBoost"])

CatBoost


,model,threshold,accuracy,precision,recall,f1_score,predicted_positive_rate
0,CatBoost,0.10,0.492458,0.340351,0.973244,0.504333,0.758651
1,CatBoost,0.15,0.568767,0.377457,0.963211,0.542373,0.677019
2,CatBoost,0.20,0.612245,0.402817,0.956522,0.566898,0.629991
3,CatBoost,0.25,0.645075,0.423831,0.939799,0.584200,0.588287
4,CatBoost,0.30,0.682343,0.451400,0.916388,0.604857,0.538598
5,CatBoost,0.35,0.707187,0.472566,0.892977,0.618056,0.501331
6,CatBoost,0.40,0.731145,0.496139,0.859532,0.629131,0.459627
7,CatBoost,0.45,0.756877,0.526427,0.832776,0.645078,0.419698
8,CatBoost,0.50,0.753327,0.523702,0.775920,0.625337,0.393079
9,CatBoost,0.55,0.773736,0.555556,0.735786,0.633094,0.351375


XGBoost


,model,threshold,accuracy,precision,recall,f1_score,predicted_positive_rate
0,XGBoost,0.10,0.635315,0.416418,0.933110,0.575851,0.594499
1,XGBoost,0.15,0.704525,0.470175,0.896321,0.616801,0.505768
2,XGBoost,0.20,0.742680,0.508982,0.852843,0.637500,0.444543
3,XGBoost,0.25,0.755102,0.525843,0.782609,0.629032,0.394854
4,XGBoost,0.30,0.767524,0.545906,0.735786,0.626781,0.357587
5,XGBoost,0.35,0.782609,0.578947,0.662207,0.617785,0.303461
6,XGBoost,0.40,0.795918,0.616162,0.612040,0.614094,0.263531
7,XGBoost,0.45,0.802130,0.646154,0.561873,0.601073,0.230701
8,XGBoost,0.50,0.803904,0.665254,0.525084,0.586916,0.209406
9,XGBoost,0.55,0.803017,0.689655,0.468227,0.557769,0.180124


## Threshold selection criteria

The optimal threshold is selected based on business-oriented considerations rather than a single metric.

For telco churn prediction:
- **Recall** is important to identify churn-risk customers.
- **Precision** is required to control unnecessary retention costs.
- **Predicted positive rate** should remain realistic for operational constraints.

Therefore, the final threshold is chosen to balance recall and precision while keeping the predicted positive rate within a reasonable range.

In [59]:
selected_thresholds = {}

MIN_THRESHOLD = 0.30
MAX_THRESHOLD = 0.70
TARGET_PPR = y_test_tree.mean()

for model_name, df in threshold_results.items():
    candidate = df[
        (df["threshold"] >= MIN_THRESHOLD) &
        (df["threshold"] <= MAX_THRESHOLD)
    ].copy()

    candidate["ppr_penalty"] = (candidate["predicted_positive_rate"] - TARGET_PPR).abs()

    candidate["score"] = (
        0.35 * candidate["recall"] +
        0.30 * candidate["precision"] +
        0.20 * candidate["f1_score"] -
        0.40 * candidate["ppr_penalty"]
    )

    best_row = candidate.sort_values("score", ascending=False).iloc[0]
    selected_thresholds[model_name] = best_row

best_threshold_df = pd.DataFrame(selected_thresholds).T.reset_index()
best_threshold_df = best_threshold_df.rename(columns={"index": "model"})
best_threshold_df

,model,model,threshold,accuracy,precision,recall,f1_score,predicted_positive_rate,ppr_penalty,score
0,CatBoost,CatBoost,0.6,0.792369,0.594203,0.685619,0.636646,0.306122,0.040686,0.529282
1,XGBoost,XGBoost,0.4,0.795918,0.616162,0.61204,0.614094,0.263531,0.001905,0.521119


Although XGBoost achieves slightly higher precision and lower predicted positive rate, CatBoost is selected due to its significantly higher recall and better overall F1-score, which are more critical in churn prediction where missing true churners is more costly than targeting some non-churners.

CatBoost is selected as the final model, and a threshold of 0.6 is chosen to balance recall and precision while reducing over-prediction, making it suitable for cost-aware churn intervention.

## Final model evaluation on test set



In [60]:
final_model_name = "CatBoost"
final_threshold = 0.6

final_model = overall_results[final_model_name]["model"]

X_test = X_test_tree
y_test = y_test_tree

y_test_proba = final_model.predict_proba(X_test)[:, 1]
y_test_pred = (y_test_proba >= final_threshold).astype(int)

final_test_metrics = {
    "test_pr_auc": average_precision_score(y_test, y_test_proba),
    "test_roc_auc": roc_auc_score(y_test, y_test_proba),
    "test_log_loss": log_loss(y_test, y_test_proba),
    "test_accuracy": accuracy_score(y_test, y_test_pred),
    "test_precision": precision_score(y_test, y_test_pred, zero_division=0),
    "test_recall": recall_score(y_test, y_test_pred, zero_division=0),
    "test_f1_score": f1_score(y_test, y_test_pred, zero_division=0),
    "test_predicted_positive_rate": y_test_pred.mean(),
}

final_test_metrics

{'test_pr_auc': 0.6583534174617631,
 'test_roc_auc': 0.8443746932237982,
 'test_log_loss': 0.4842078212526892,
 'test_accuracy': 0.7750177430801988,
 'test_precision': 0.5631929046563193,
 'test_recall': 0.679144385026738,
 'test_f1_score': 0.6157575757575757,
 'test_predicted_positive_rate': np.float64(0.32008516678495386)}

In [61]:
print("Final Model:", final_model_name)
print("Threshold:", final_threshold)
print("-" * 40)

for k, v in final_test_metrics.items():
    print(f"{k}: {v:.4f}")

Final Model: CatBoost
Threshold: 0.6
----------------------------------------
test_pr_auc: 0.6584
test_roc_auc: 0.8444
test_log_loss: 0.4842
test_accuracy: 0.7750
test_precision: 0.5632
test_recall: 0.6791
test_f1_score: 0.6158
test_predicted_positive_rate: 0.3201


Final model: CatBoost (threshold = 0.6) delivers strong discriminative performance (ROC-AUC 0.84, PR-AUC 0.66) with a balanced precision–recall trade-off (precision 0.56, recall 0.68) and a controlled prediction rate (~32%), making it suitable for practical, cost-aware churn prediction.

# Export final config

In [62]:
from pathlib import Path
import yaml

CONFIGS_DIR = Path("../configs")
CONFIGS_DIR.mkdir(parents=True, exist_ok=True)

config = {
    "project": "Telco_Churn_Model",  # hoặc tên bạn muốn trên MLflow
    "best_model_overall": final_model_name,  # phải khớp với train.py
    "threshold": float(final_threshold),
    "hyperparameters": overall_results[final_model_name]["params"],
}

with open(CONFIGS_DIR / "best_model.yaml", "w") as f:
    yaml.dump(config, f, sort_keys=False)

print("Saved config to ../configs/best_model.yaml")

Saved config to ../configs/best_model.yaml
